pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go

```diff
@@ -0,0 +1,397 @@
+
+		// external is the read-only view of the spanconfig.Store maintained by
+		// the KVSubscriber. It's usually an opaque alias to an internal
+		// spanconfig.Store, but when re-subscribing, a new spanconfig.Store is
+		// populated while the existing StoreReader remains static. Once the new
+		// spanconfig.Store is sufficiently caught up, the external view is
+		// aliased to the new spanconfig.Store and the old
+		// spanconfig.StoreReader is discarded. See the type-level comment for
+		// more details.
+		external spanconfig.StoreReader
+	}
+
+	lastFrontierTS hlc.Timestamp
+}
+
+var _ spanconfig.KVSubscriber = &KVSubscriber{}
+
+// spanConfigurationsTableRowSize is an estimate of the size of a single row in
+// the system.span_configurations table (size of start/end key, and size of a
+// marshaled span config proto). The value used here was pulled out of thin air
+// -- it only serves to coarsely limit how large the KVSubscriber's underlying
+// rangefeed buffer can get.
+const spanConfigurationsTableRowSize = 5 << 10 // 5 KB
+
+// New instantiates a KVSubscriber.
+func New(
+	stopper *stop.Stopper,
+	db *kv.DB,
+	clock *hlc.Clock,
+	rangeFeedFactory *rangefeed.Factory,
+	spanConfigurationsTableID uint32,
+	bufferMemLimit int64,
+	fallback roachpb.SpanConfig,
+	knobs *spanconfig.TestingKnobs,
+) *KVSubscriber {
+	spanConfigTableStart := keys.SystemSQLCodec.IndexPrefix(
+		spanConfigurationsTableID,
+		keys.SpanConfigurationsTablePrimaryKeyIndexID,
+	)
+	spanConfigTableSpan := roachpb.Span{
+		Key:    spanConfigTableStart,
+		EndKey: spanConfigTableStart.PrefixEnd(),
+	}
+	spanConfigStore := spanconfigstore.New(fallback)
+	if knobs == nil {
+		knobs = &spanconfig.TestingKnobs{}
+	}
+	s := &KVSubscriber{
+		stopper:             stopper,
+		db:                  db,
+		clock:               clock,
+		bufferMemLimit:      bufferMemLimit,
+		rangefeedFactory:    rangeFeedFactory,
+		spanConfigTableSpan: spanConfigTableSpan,
+		fallback:            fallback,
+		knobs:               knobs,
+		decoder:             newSpanConfigDecoder(),
+	}
+	s.mu.external = spanConfigStore
+	return s
+}
+
+// Subscribe establishes a rangefeed over the global store of span configs. This
+// is a blocking operation, returning (read: unsubscribing) only when the
+// surrounding stopper is stopped or when an error occurs. The installed handler
+// is invoked in this single main thread. When unsubscribed, the exposed
+// spanconfig.StoreReader continues to be readable (though it's no longer
+// incrementally maintained -- the view gets progressively staler overtime).
+//
+// It's expected that callers will re-resubscribe in the face of retryable
+// errors. Any existing handler is kept intact and will be notified when the
+// rangefeed is re-established. After a successful re-subscription, the exposed
+// StoreReader will be up-to-date and continue to be incrementally maintained.
+func (s *KVSubscriber) Subscribe(ctx context.Context) error {
```

Comment:
all around, I find this model to be surprising. The pattern that I'd expect would be for there to be a `Start` method that does effectively this subscribe loop plus a retry loop around it. Then, I'd expect that the `OnSpanConfigUpdate` could be called at any point, before or after `Start` and that upon `Start` or an internal restart we'd notify all of the subscribers. 

I don't see much benefit putting the retry loop in the server package as opposed to this package.

Reply:
Great suggestion, moved the retry loop inside the package.